### ISSUE TITLE
Quiver as a library

### ISSUE THREAD

**Issue by suhr** [varkor/quiver#76]
Though Quiver does not support export in SVG, it can display diagrams in browser. It would be great to have Quiver as a library, so it could be used (as a React component) in articles written using [Idyll](https://idyll-lang.org).

**Comment by varkor** [comment]
Could you explain your use case, so it's clearer what would need to be supported in **quiver** to facilitate this?

**Comment by suhr** [comment]
Basically, I want to render a diagram inside a DOM element. Just like KaTeX renders formulas: https://katex.org/docs/api.html

Use case: this allows to embed Quiver as a custom component into a page. So I can have nice commutative diagrams on a page without having to make PNG screenshots of diagrams inside the browser.

**Comment by varkor** [comment]
@AviCraimer is currently working on the ability to embed **quiver** in an `<iframe>` without the usual editing interaction and UI elements. Do you think this would suffice for your purposes?

**Comment by suhr** [comment]
iframe is rather awkward, but some js will be probably enough to extract the diagram from it.

**Comment by enjoysmath** [comment]
I'm also having trouble making a viewer-only out of quiver for the project:

https://github.com/enjoysmath/QuiverDatabase

I tried putting an 
        ``` if (quiver_edit_mode) { ... }```
around certain parts (took forever) and it didn't work.  I've also tried html2canvas and the arrows come out improperly positioned.  It would be nice to have a view of a diagram without taking an image screenshot and have to deal with image hosting etc.  So right now, I'm limited to showing the diagram name only in any diagram result list (say for searching).  

In other words, I would also like to see readonly mode for quiver.

**Comment by enjoysmath** [comment]
I saw an embed button patch to the repo somewhere.  @varkor  what branch would I check out to get that code?

**Comment by varkor** [comment]
@enjoysmath: take a look at [this help page](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).

**Comment by davidson16807** [comment]
I'm also debating whether I should build off quiver for my own ideas.  I namely want to repurpose the user interface to organize software projects, so if I ctrl+click on an arrow or object then I could jump to a file in an IDE, or possibly color arrows to reflect whether unit tests pass (BTW, I'm very glad that colored arrows are already supported!). This might require either embedding the code in an electron app or in a plugin for a text editor like vscode. It also requires some way to attach arbitrary metadata to objects or arrows, ideally without having to modify existing classes, maybe as simple as a dictionary that maps the ids for arrows or objects to arbitrary metadata. 

The only reason I think I can do this with quiver right now is because the code base is rather small and katex is the only dependency, so I'm worried that won't be the case in the future as additional features for a library are implemented. Would it be better to create a lite version, instead, something like a fairly static fork that people could ease into without requiring them to work within known use cases?

**Comment by varkor** [comment]
I can't comment on the applicability for a software project organisation tool, but I would think it should be straightforward to make an Electron app. The data structure for diagrams makes adding metadata very easy too.

> The only reason I think I can do this with quiver right now is because the code base is rather small and katex is the only dependency, so I'm worried that won't be the case in the future as additional features for a library are implemented.

I don't see any further dependencies being added to **quiver**. LaTeX rendering is something that needs to be done by an external library, but I can't think of many other features that would necessitate an external dependency.

**Comment by varkor** [comment]
@suhr: **quiver** now supports exporting embeddable diagrams as HTML. Is this sufficient for your purposes?

**Comment by suhr** [comment]
It probably is, let me check it out.

**Comment by suhr** [comment]
Yeah, it works. Some notes though:

1. It takes some time to load instead of loading instantly
2. It would be great to be able to hide the quiver logo or at least make it smaller
3. Being an iframe, it requires an internet connection to load diagrams

**Comment by varkor** [comment]
To be honest, I feel the right solution to each of these problems is to get around to implementing #8, and use MathJax for label rendering for now, unless you want interaction in the exported diagrams (e.g. like zooming or panning).

**Comment by varkor** [comment]
I'm going to close this in favour of #8, which I think would solve your problems. Unfortunately, #8 is blocked right now, but there may be ways to work around it.
